### PR TITLE
Add style guide

### PR DESCRIPTION
**Pull request by Phlip79**
# What does this PR do ?

Adding a style guide. This style guide can be used by humans but the primary purpose is to guide AI. Inspired by MBridge's [CODING_GUIDELINES.md](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CODING_GUIDELINES.md).

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

## Summary by CodeRabbit

* **Documentation**
  * Added comprehensive style guide documenting coding conventions, naming standards, import organization, documentation practices, testing guidelines, and copyright requirements for the project.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by Phlip79** [comment]
@coderabbitai review

**Review by ericharper — APPROVED** [review]

**Comment by sbhavani** [comment]
@Phlip79 looks great! if the main purpose is for AI, could we place this in an `AGENTS.md` at the root? that would be easiest for discovery by vscode/cursor.

File: style-guide.md [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+   - upper snake_case: `MY_CONSTANT = ...`
+
+#### Identifier Guidelines
+
+1. Avoid shadowing variables declared in an outer scope.
+2. Initialize all externally visible members of a class in the constructor.
+
+### Imports
+
+Organize imports in the following order, separated by blank lines:
```
Comment (jaredcasper) [inline style-guide.md:55]: Maybe mention isort?

**Comment by Phlip79** [comment]
> looks great! if the main purpose is for AI, could we place this in an AGENTS.md at the root? that would be easiest for discovery by vscode/cursor.

Primary purpose is AI, secondary purpose is humans. But that doesn't mean we should forget about the humans :)

File: style-guide.md [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+### Documentation Requirements
+
+**Important**: All new key features (e.g., enabling a new model, enabling a new parallelism strategy) must include documentation updates. This should:
+
+- Explain the motivation and purpose of the feature
+- Outline the technical approach and architecture
+- Provide clear usage examples and instructions for users
+- Document internal implementation details where appropriate
+
+## Testing Guidelines
```
Comment (jaredcasper) [inline style-guide.md:261]: Link to existing test code/guidelines?

File: style-guide.md [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+    """Test end-to-end model loading."""
+    ...
+```
+
+## NVIDIA Copyright
+
+Add the following NVIDIA copyright header to all Python files and shell scripts. The header should appear at the top of the file:
+
+```python
+# Copyright (c) 2026, NVIDIA CORPORATION.  All rights reserved.
```
Comment (jaredcasper) [inline style-guide.md]: We don't put the whole license in each file, just the copyright.

File: style-guide.md [hunk 4]
```diff
@@ -0,0 +1,321 @@
+# Megatron-LM Coding Guidelines
+
+Note: This repository is Python-first. Prefer the Python guidelines in this document.
+
+## Style Guides We Follow
+
+- Python: [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html)
```
Comment (jaredcasper) [inline style-guide.md:9]: Do we have consensus that this the style guide we should follow?

File: style-guide.md [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+- Shell: [Google Shell Style Guide](https://google.github.io/styleguide/shellguide.html)
+
+## uv Guidelines
+
+### Use uv run instead of python
+
+Use `uv run` to execute scripts, rather than activating a virtual environment and calling `python` directly.
```
Comment (jaredcasper) [inline style-guide.md]: Why?

File: style-guide.md [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+    ... # Not a file-like object, do something else
+else:
+    f.seek(0)
+    f.read()
+```
+
+### Type Hints
+
+1. Use type hints for function arguments and return types.
+2. Use `T | None` for nullable types (not `Optional[T]`).
```
Comment (jaredcasper) [inline style-guide.md]: This is not what most of the code currently does.

File: style-guide.md [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+3. Use `X | Y` for union types (not `Union[X, Y]`).
```
Comment (jaredcasper) [inline style-guide.md]: Same here, I believe we use Union mostly.

File: style-guide.md [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+) -> dict[str, torch.Tensor]:
+    """Convert weights, accepting either a single tensor or a dict."""
+    ...
+```
+
+### Configuration and Dataclasses
+
+1. Use `dataclasses` or `NamedTuple` for configuration objects.
+2. Be explicit about required vs optional fields.
+3. Do not add arbitrary defaults for configs; be as explicit as possible.
```
Comment (jaredcasper) [inline style-guide.md:195]: I wish we followed this more in the code.

File: style-guide.md [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+Example:
+
+```python
+from dataclasses import dataclass
+
+@dataclass
+class ModelConfig:
+    """Configuration for model architecture."""
+
+    hidden_size: int
```
Comment (jaredcasper) [inline style-guide.md]: We want docstrings under each element, maybe should put that here (with fewer examples maybe)

**Review by jaredcasper — COMMENTED** [review]
Maybe also mentioned things like avoiding **kwargs except in very specific cases, etc.? But I guess we can add those later.

File: style-guide.md (hunk 6 above)
Comment (Phlip79) [reply style-guide.md]: I want to establish a guideline for future changes. I think it's ok if existing code does not conform. What do you think?

File: style-guide.md (hunk 9 above)
Comment (Phlip79) [reply style-guide.md]: Done.

File: style-guide.md (hunk 3 above)
Comment (Phlip79) [reply style-guide.md]: Updated.

File: style-guide.md (hunk 1 above)
Comment (Phlip79) [reply style-guide.md:55]: Done.

**Comment by skyw** [comment]
I strongly support having guidelines and enforce it. But the guideline needs to be unambiguous. 

For example, in the python guideline section
```
from transformers import PreTrainedModel
```
Importing class is prohibited by Google style (https://google.github.io/styleguide/pyguide.html#22-imports) `isort` can also easily break google style.

PyTorch (assume meta also) does allow these type of imports. either way, just need a clear choice and agreement.

File: style-guide.md (hunk 7 above)
Comment (skyw) [reply style-guide.md]: Python typing syntax evolved a lot over time. in addition to "|", now capitalized List, Tuple, etc. are not needed. Need guideline for those as well.

**Comment by Phlip79** [comment]
> Importing class is prohibited by Google style (https://google.github.io/styleguide/pyguide.html#22-imports) `isort` can also easily break google style.

Great point. Updated to mention `isort --profile google`. This doc takes precedence over the Google style guide. I just added "If there is a conflict between this file and the Google style guides, prioritize this file."

File: style-guide.md (hunk 5 above)
Comment (Phlip79) [reply style-guide.md]: Makes life a lot easier when running locally and debugging in my experience.

Comment (jaredcasper) [reply style-guide.md]: But that's more of a personal workflow preference, not a style that should be asked of everyone?

File: style-guide.md (hunk 6 above)
Comment (jaredcasper) [reply style-guide.md]: I personally think `Optional[T]` is more clear, it indicates that sending `None` means the argument is not provided vs an argument where `None` as a more distinct meaning. If this is the guidelines for Python in general going forward we can match the community, but if this is just us deciding I would vote for `Optional[T]`

**Comment by jaredcasper** [comment]
> Great point. Updated to mention `isort --profile google`. This doc takes precedence over the Google style guide. I just added "If there is a conflict between this file and the Google style guides, prioritize this file."

We don't currently use `isort --profile google` though? Are you proposing we switch import ordering across the whole codebase?

File: style-guide.md (hunk 6 above)
Comment (skyw) [reply style-guide.md]: Python's "official" opinion, https://typing.python.org/en/latest/reference/best_practices.html#shorthand-syntax

Of course final decision still depend on project. above is just a recommendation for new code.

File: style-guide.md (hunk 4 above)
Comment (janEbert) [reply style-guide.md:9]: On the MLM Wiki on GitLab, this is also suggested. I don't know how that decision came to be, though.

File: style-guide.md (hunk 5 above)
Comment (janEbert) [reply style-guide.md]: This is what's used to create the container in the `docker/Dockerfile.ci.dev` and for running tests in (some of) the convenience helper scripts. The NGC PyTorch container also includes `uv` already.

Seems to me like it's good to recommend it, since it has already become the unofficial, but "productionized" way to run already, due to its use in the CI. The rest of the docs should be similarly updated to reflect this, though. At the moment, it's only mentioned/used in the documentation for documentation generation instructions.

I would actually vote for asking this of everyone so that it's easier to have a common setup. We can more easily assume that the setup respects the same settings for `pyproject.toml` that we and the CI would also use. It's also easier to only maintain one set of usage instructions.

I also understand possible reluctance toward not simply working with standard tooling. However, the users that would prefer working like that could still do it; it would just not be "officially supported"/recommended.

File: style-guide.md [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+
+3. Functions and Methods
+   - snake_case: `def my_awesome_function():`
+
+4. Local Variables
+   - snake_case: `my_variable = ...`
+   - prefix `k` for variable names that start with a number: `k_99th_percentile = ...`
+
+5. Global Variables
+   - upper snake_case and prefix `G`: `G_MY_GLOBAL = ...`
```
Comment (janEbert) [inline style-guide.md]: The codebase doesn't follow this suggestion at this point and my personal preference would also speak against this.

File: style-guide.md [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+    x, y = args
+    return dict(**locals())
+```
+
+Do:
+
+```python
+def make_complex(x, y):
+    return {"x": x, "y": y}
+```
```
Comment (janEbert) [inline style-guide.md]: This seems extremely specific to mention; I would prefer moving this section further down due to its specificity.

File: style-guide.md (hunk 5 above)
Comment (sbhavani) [reply style-guide.md]: I'd vote for leaving it as pip until we update documentation and have tests for uv installation experience. I know TE has some known issues installing using uv's virtual env

File: style-guide.md (hunk 10 above)
Comment (Phlip79) [reply style-guide.md]: Removed prefix `G` mention.

File: style-guide.md (hunk 11 above)
Comment (Phlip79) [reply style-guide.md]: Moved, PTAL

**Comment by Phlip79** [comment]
I've removed references to `isort`.

File: style-guide.md (hunk 5 above)
Comment (Phlip79) [reply style-guide.md]: Removed uv mention.

Comment (janEbert) [reply style-guide.md]: I agree with having the documentation as the main reference to follow.

But isn't the recommendation to use the code with the NGC PyTorch container anyway, which has TE pre-installed? Do we need to worry about potential `uv` issues in that case? Since all CI images use `uv` for MLM setup, I would considered its installation experience tested, but maybe the non-NGC case also needs to be covered?

File: style-guide.md (hunk 11 above)
Comment (janEbert) [reply style-guide.md]: Thanks!

**Comment by Phlip79** [comment]
Any other comments before I merge this PR? It's been long overdue.

File: style-guide.md [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+    f.read()
+```
+
+### Type Hints
+
+1. Use type hints for function arguments and return types.
+2. Use `Optional[T]` for nullable types.
+3. Use `Union[X, Y]` for union types.
+4. Use `TypeVar` for generic type parameters.
+5. Use `typing` generics (`List`, `Dict`, `Tuple`) for type annotations. Built-in generics (`list`, `dict`) are also acceptable.
```
Comment (janEbert) [inline style-guide.md:167]: Should be updated to use Python 3.12 syntax.

File: style-guide.md [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+
+1. Use **double quotes** for strings.
+
+### Comments
+
+1. For interfaces that may be used outside a file, prefer docstrings over comments.
+2. Comments should be reserved for code within a function, or interfaces that are local to a file.
+3. If a piece of code is commented out, there should be a comment around that piece of code describing its usage and why it's commented out. Otherwise that is a debug comment and it should be removed before merging.
+
+### Docstring Syntax
```
Comment (janEbert) [inline style-guide.md:87]: We badly need a section on docstring format here. E.g., should docstrings be MyST like the rest of the codebase (and what's used by our documentation generator), reST like AIs often default to, or something else? I understand the Google style guide doesn't have a decision on this and it'd be useful to have consistency.

File: style-guide.md [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+   - snake_case: `def my_awesome_function():`
+
+4. Local Variables
+   - snake_case: `my_variable = ...`
+
+5. Global Variables
+   - upper snake_case: `MY_GLOBAL = ...`
+
+6. Constants
+   - upper snake_case: `MY_CONSTANT = ...`
```
Comment (Phlip79) [inline style-guide.md:46]: Is this necessary? Looks to be duplicate.

File: style-guide.md (hunk 1 above)
Comment (Phlip79) [reply style-guide.md:55]: Need to add mention of isort again. It is an override of google style guide.

File: style-guide.md [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+
+#### Classes and Functions
+
+Use the [Google style](https://google.github.io/styleguide/pyguide.html), which can be parsed by Sphinx.
+
+Example:
+
+```python
+def sharded_state_dict(
+    self, prefix: str = '', sharded_offsets: tuple = (), metadata: Optional[Dict] = None
```
Comment (Phlip79) [inline style-guide.md:97]: Update to use double quotes.

File: style-guide.md [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+        sharded_offsets (tuple): PP related offsets, expected to be empty at this module level.
+        metadata (Optional[Dict]): metadata controlling sharded state dict creation.
+
+    Returns:
+        ShardedStateDict: sharded state dict for the GPTModel
+    """
+    ...
+```
+
+### Error Handling
```
Comment (Phlip79) [inline style-guide.md:115]: Remove this section as it's duplicate.

File: style-guide.md [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+try:
+    f.seek  # Do not call to minimize chance of unrelated failure
+except AttributeError:
+    ... # Not a file-like object, do something else
+else:
+    f.seek(0)
+    f.read()
+```
+
+### Type Hints
```
Comment (Phlip79) [inline style-guide.md:161]: Check this whole section for duplication

File: style-guide.md [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+
+1. Use type hints for function arguments and return types.
+2. Use `Optional[T]` for nullable types.
+3. Use `Union[X, Y]` for union types.
```
Comment (Phlip79) [inline style-guide.md:165]: Remove these.

File: style-guide.md [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+    including those in ModelParallelConfig.
+    """
+
+    num_layers: int = field(default=0, metadata={"argparse_meta": {"default": None}})
+    """Number of transformer layers in a transformer block."""
+
+    mtp_num_layers: Optional[int] = None
+    """Number of Multi-Token Prediction (MTP) Layers."""
+
+    mtp_loss_scaling_factor: Optional[float] = 0.1
```
Comment (Phlip79) [inline style-guide.md:217]: This is a bad example.

File: style-guide.md [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+    ...
+
+def convert_weights(
+    weights: Union[torch.Tensor, Dict[str, torch.Tensor]],
+) -> Dict[str, torch.Tensor]:
+    """Convert weights, accepting either a single tensor or a dict."""
+    ...
+```
+
+### Configuration and Dataclasses
```
Comment (Phlip79) [inline style-guide.md:191]: Add a bit about making dataclasses specific, not overly generic
